Source/DafnyCore/AST/Formatting.dfy

```diff
@@ -0,0 +1,157 @@
+
+  trait {:extern "IToken"} {:compile false} IToken {
+    var val : CsString
+    var LeadingTrivia: CsString
+    var TrailingTrivia: CsString
+    var Next: IToken?
+    ghost var allTokens: seq<IToken>
+
+    ghost predicate Valid() reads this, allTokens decreases |allTokens| {
+      && |allTokens| > 0 && allTokens[0] == this
+      && (|| (|allTokens| == 1 && Next == null)
+          || (&& |allTokens| > 1
+              && Next == allTokens[1]
+              && Next.allTokens == allTokens[1..]
+              && Next.Valid())
+         )
+    }
+    lemma AlltokenSpec(i: int)
+      requires Valid()
+      decreases |allTokens|
+      requires 0 <= i < |allTokens|
+      ensures allTokens == allTokens[..i] + allTokens[i].allTokens
+    {
+      if i == 0 {
+      } else {
+        this.Next.AlltokenSpec(i - 1);
+      }
+    }
+    lemma TokenNextIsIPlus1(middle: IToken, i: int)
+      requires Valid()
+      requires 0 <= i < |allTokens|
+      requires allTokens[i] == middle;
+      requires middle.Next != null
+      ensures 0 <= i + 1 < |allTokens| && allTokens[i+1] == middle.Next
+    {
+      if i == 0 {
+      } else {
+        this.Next.TokenNextIsIPlus1(middle, i - 1);
+      }
+    }
+
+    lemma TokenNextIsNullImpliesLastToken(middle: IToken, i: int)
+      requires middle.Valid() && this.Valid()
+      requires 0 <= i < |allTokens|
+      requires middle == allTokens[i]
+      requires middle.Next == null
+      ensures i == |allTokens|-1
+    {
+      if Next == null {
+      } else {
+        Next.TokenNextIsNullImpliesLastToken(middle, i-1);
+      }
+    }
+  }
+  class {:extern "HelperString"} {:compile false} HelperString {
+    static predicate FinishesByNewline(input: CsString)
+  }
+}
+module {:extern "Microsoft"} {:options "-functionSyntax:4"}  Microsoft {
+  module {:extern "Dafny"} Dafny {
+    module {:extern "Formatting"} Formatting {
+      import opened MicrosoftDafny
+      import opened System
+
+      const {:extern "System", "String.Empty"} CsStringEmpty: CsString;
+
+      trait IIndentationFormatter {
+        function Reindent(token: IToken, trailingTrivia: bool, precededByNewline: bool, indentation: CsString, lastIndentation: CsString): CsString
+        // Returns -1 if no indentation is set
+        method GetIndentation(token: IToken, currentIndentation: CsString)
+          returns (
+            indentationBefore: CsString,
+            indentationBeforeSet: bool,
+            lastIndentation: CsString,
+            lastIndentationSet: bool,
+            indentationAfter: CsString,
+            indentationAfterSet: bool)
+          requires token.Valid()
+          ensures token.allTokens == old(token.allTokens)
+      }
+
+      lemma IsAllocated<T>(x: T)
+        ensures allocated(x) {
+      }
+
+      /** Prints the entire program while fixing identation, based on a map */
+      method printSourceReindent(firstToken: IToken, reindent: IIndentationFormatter) returns (s: CsString)
```

Comment:
1. Do I understand correctly that "entire program" refers to `firstToken.allTokens`?
2. What is meant by "based on a map"? I can't seem to find a map used anywhere in the body of this method.

Reply:
> Do I understand correctly that "entire program" refers to firstToken.allTokens?
> What is meant by "based on a map"? I can't seem to find a map used anywhere in the body of this method.

1. Yes, the entire program's token is obtained by traversing the tokens starting at `firstToken`, the first token of the program. `allTokens` is a ghost variable that only keeps track of all the successors of `firstToken`.
2. I changed the comment to:

```dafny
/** Prints the entire program while fixing identation, based on
          1) indentation information provided by the IIndentationFormatter reindent
          2) Reindentation algorithm provided by the same reindent */
```